src/qt3utils/datagenerators/plescanner.py

```diff
@@ -0,0 +1,266 @@
+        if self.wavelength_controller:
+            self.go_to(wl_point=self._start)
+
+    def close(self) -> None:
+        """
+        close connection to daq.
+        """
+        pass
+
+    def set_scan_range(self, _start: float, _end: float) -> None:
+        """
+        Set the wavelength controller's minimum and maximum for the PLE scan
+        """
+        if self.wavelength_controller:
+            self.wavelength_controller.check_allowed_limits(_start)
+            self.wavelength_controller.check_allowed_limits(_end)
+        self._start = _start
+        self._end = _end
+
+    def get_scan_range(self) -> tuple[float, float]:
+        """
+        Returns a tuple of the full scan range
+        :return: _start, _end
+        """
+        return self._start, self._end,
+
+    def get_completed_scan_range(self) -> tuple[float, float, float]:
+        """
+        Returns a tuple of the scan range that has been completed
+        :return: starting and ending points of wavelength contrl, as well as current point in run
+        """
+        return self._start, self._end, self.current_wl_point
+
+    def still_scanning(self) -> bool:
+        """
+        Return boolean to determine of the scan should continue
+        returning True means the scan should continue
+        returning False means the scan should stop
+        """
+        if self.running == False:  # this allows external process to stop scan
+            return False
+
+        if self.current_frame < self.tmax:  # stops scan when reaches final position
+            return True
+        else:
+            self.running = False
+            return False
+
+    def increment_wl(self) -> None:
+        """
+        Move the wavelength control one _step_size towards maximum
+        """
+        self.current_wl_point += self._step_size
+        if self.wavelength_controller and self.current_wl_point <= self._end:
+            try:
+                self.wavelength_controller.go_to(wl_point=self.current_wl_point)
+            except ValueError as e:
+                logger.info(f'out of range\n\n{e}')
+
+    def go_to(self, wl_point: float) -> None:
+        """
+        Set the wavelength controller to desired point
+        """
+        if self.wavelength_controller and wl_point <= self._end and wl_point >= self._start:
+            try:
+                self.wavelength_controller.go_to(wl_point=wl_point)
+            except ValueError as e:
+                logger.info(f'out of range\n\n{e}')
+
+    def scan_wavelengths(self) -> None:
+        """
+        Scans the wavelengths from v_start to v_end in steps of step_size.
+        Records the output from the readers into a list of dictionaries
+        """
+        self.outputs.append(self.scan_axis(self.axis_name, self._start, self._end, self._step_size))
+        self.current_frame = self.current_frame + 1
+
+    def scan_axis(self, axis: str, min: float, max: float, step_size: float) -> list:
```

Comment:
Is it possible to allow for aborting the scan mid way through the axis scan? Some single axis scans may take from a few seconds to 20 minutes, so it is best if we can stop the ple midway through the axis scan

Reply:
Yes. There is a "Stop scan" button in applications/qt3ple/main.py which will stop the background thread sweeping the wavelength control, but retain the connections to the devices and then you can either save off what you have collected or start another scan.